client/landing/gutenboarding/components/domain-picker-button/index.tsx

```diff
@@ -76,20 +76,28 @@ const DomainPickerButton: FunctionComponent< Props > = ( {
 	const [ isDomainPopoverVisible, setDomainPopoverVisibility ] = useState( false );
 	const [ isDomainModalVisible, setDomainModalVisibility ] = useState( false );
 
+	const handlePopoverOpen = () => {
+		recordEnterModal( 'DomainPickerPopover' );
+	};
+
 	const handlePopoverClose = ( e?: React.FocusEvent ) => {
 		// Don't collide with button toggling
 		if ( e?.relatedTarget === buttonRef.current ) {
 			return;
 		}
 		setDomainPopoverVisibility( false );
+		recordCloseModal( 'DomainPickerPopover', {
+			selected_domain: getSelectedDomain()?.domain_name,
```

Comment:
Could the observed bug be related to this line?

https://github.com/Automattic/wp-calypso/blob/28e988feee80ce9e0ed1f595d190435e67c72b21/client/landing/gutenboarding/components/domain-picker-button/index.tsx#L69

I don't believe returning selectors from a `wp-data` is valid behavior… the selectors are meant to be invoked inside the `useSelect` callback and return a value. There's caching in place in `useSelect` and I suspect that return a selector (possibly an unexpectedly stable value) is causing the system to behave in unexpected ways.

Reply:
To elaborate, I'd expect that to be more like:

```js
const selectedDomain = useSelect( ( select ) => select( STORE_KEY ).getSelectedDomain() );
```

and then we'd use the value directly here rather than invoking the stored selector:

```suggestion
			selected_domain: selectedDomain?.domain_name,
```

Reply:
Good one!

Reply:
Okay so the track event fires before the updated selected domain gets propagated down from the parent down to the child component, that's why the suggestion didn't work.

There are surely ways around this but I'm going to yank the firing of events out from DomainPickerButton and do it separately in DomainPickerPopover and DomainPickerModal as suggested by @razvanpapadopol here https://github.com/Automattic/wp-calypso/pull/42659#discussion_r430891845.

Reply:
Okay some updates. Due to the fact that there are 3 ways a popover gets closed (clicking outside, pressing ESC key, clicking on close button) and 3 ways a modal gets closed (clicking Confirm, clicking Cancel & clicking ESC key) where the closing event is handled either the parent component (popover/modal component) or the domain picker component itself, I did not invest the time into checking what values it holds when triggered by whom, so I'm resorting back to `selected_domain: getSelectedDomain()?.domain_name` for now which surely returns the currently selected domain name from store.

Reply:
I won't stand in the way of changes that are demonstrably working, but returning the selector fro `useSelect` is unorthodox, at least.

[The docs](https://github.com/WordPress/gutenberg/tree/master/packages/data#useSelect) return the value returned by the selector.

🤔  This has me thinking. If what we want here is to call a selector at an arbitrary time, why aren't we using [`select`](https://github.com/WordPress/gutenberg/tree/master/packages/data#select) directly? `useSelect` is tied to renders, but if we don't need render updated values, but rather we need to get a selected value e.g. when a user clicks, then perhaps we should be calling `select` directly in the click handler.

Reply:
> If what we want here is to call a selector at an arbitrary time, why aren't we using select directly?
Ah yes! Makes sense! Let me try calling `select( STORE_KEY ).getSelectedDomain()` directly!

Reply:
Works well! I'm pretty happy with this now. :) https://github.com/Automattic/wp-calypso/pull/42659/commits/10008fee61498fcd1acbe91072308adee545fdcd

Reply:
That's better 🙂